sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/credential/PopTokenAuthenticationPolicy.java

```diff
@@ -0,0 +1,199 @@
+        }
+        HttpPipelineNextPolicy nextPolicy = next.clone();
+
+        return authorizeRequest(context).then(Mono.defer(next::process)).flatMap(httpResponse -> {
+            String authHeader = httpResponse.getHeaderValue(HttpHeaderName.WWW_AUTHENTICATE);
+            if (httpResponse.getStatusCode() == 401 && authHeader != null) {
+                return authorizeRequestOnChallenge(context, httpResponse).flatMap(authorized -> {
+                    if (authorized) {
+                        // body needs to be closed or read to the end to release the connection
+                        httpResponse.close();
+                        return nextPolicy.process();
+                    } else {
+                        return Mono.just(httpResponse);
+                    }
+                });
+            } else if (authHeader != null) {
+                popNonce = AuthorizationChallengeParser.getChallengeParameterFromResponse(httpResponse, "PoP", "nonce");
+            }
+            return Mono.just(httpResponse);
+        });
+    }
+
+    /**
+     * Processes the request synchronously.
+     * @param context The context of the request.
+     * @param next The next policy in the pipeline.
+     * @return The {@link HttpResponse}.
+     */
+    public HttpResponse processSync(HttpPipelineCallContext context, HttpPipelineNextSyncPolicy next) {
+        if (!"https".equals(context.getHttpRequest().getUrl().getProtocol())) {
+            throw LOGGER.logExceptionAsError(
+                new RuntimeException("token credentials require a URL using the HTTPS protocol scheme"));
+        } else {
+            HttpPipelineNextSyncPolicy nextPolicy = next.clone();
+            this.authorizeRequestSync(context);
+            HttpResponse httpResponse = next.processSync();
+            String authHeader = httpResponse.getHeaderValue(HttpHeaderName.WWW_AUTHENTICATE);
+            if (httpResponse.getStatusCode() == 401 && authHeader != null) {
+                if (this.authorizeRequestOnChallengeSync(context, httpResponse)) {
+                    httpResponse.close();
+                    return nextPolicy.processSync();
+                } else {
+                    return httpResponse;
+                }
+            } else if (authHeader != null) {
+                popNonce = AuthorizationChallengeParser.getChallengeParameterFromResponse(httpResponse, "PoP", "nonce");
+                return httpResponse;
+            } else {
+                return httpResponse;
+            }
+        }
+    }
+
+    private Mono<Void> setAuthorizationHeaderHelper(HttpPipelineCallContext context, boolean checkToForceFetchToken) {
+        if (!"https".equals(context.getHttpRequest().getUrl().getProtocol())) {
+            throw LOGGER.logExceptionAsError(new RuntimeException(
+                "Proof of possession token authentication is not permitted for non TLS-protected (HTTPS) endpoints."));
```

Comment:
Not sure using a different string than we use above is worth it since we already require https anyway?

Reply:
can be removed yes, will remove it next commit.

Reply:
on second look, authorizeRequest and authorizeRequestOnChallenge public APIs directly call this method too. 
So, need to retain this check here. 
Aligned the error message to be same in all spots.